File: KnightTour/knight.tour.open.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n, m, cnt;
vector<vector<int>> graph;
vector<int> mark;

void make_graph() {
    int imov[] = {-2, -1, 1, 2,  2,  1, -1, -2};
    int jmov[] = { 1,  2, 2, 1, -1, -2, -2, -1};
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            for (int k = 0; k < 8; k++) {
                int ni = i + imov[k];
                int nj = j + jmov[k];
                if (ni >= 0 && nj >= 0 && ni < n && nj < m)
                    graph[i*m + j].push_back(ni*m + nj);
            }
}

void tour(int kth, int v) {
    if (kth == n * m) {
        cnt++;
        // cout << "solution #" << cnt << ":" << endl;
        // for (int i = 0; i < n; i++) {
        //     for (int j = 0; j < m; j++)
        //         printf("%2d ", mark[i*m + j]);
        //     cout << endl;
        // }
    } else {
        for (int u: graph[v])
            if (mark[u] == 0) {
                mark[u] = kth + 1;
                tour(kth + 1, u);
                mark[u] = 0;
            }
    }
}

int main() {
    cin >> n >> m;
    graph.resize(n * m, vector<int>(0));
    make_graph();
    for (int v = 0; v < n * m; v++) {
        cout << v << ": ";
        for (int j = 0; j < graph[v].size(); j++)
            cout << graph[v][j] << " ";
        cout << endl;
    }

    int total = 0;
    for (int s = 0; s < n * m; s++) {
        mark.resize(n * m, 0);
        fill(mark.begin(), mark.end(), 0);
        mark[s] = 1; // mark starting vertex
        cnt = 0;
        tour(1, s);
        if (cnt > 0)
            cout << s << ": " << cnt << endl;
        total += cnt;
    }
    cout << "total = " << total << endl;
}

File: KnightTour/knight.tour.open.py
def make_graph(n, m):
    imove = [-2, -1, 1, 2,  2,  1, -1, -2]
    jmove = [ 1,  2, 2, 1, -1, -2, -2, -1]
    graph = {i:[] for i in range(n * m)}
    for i in range(n):
        for j in range(m):
            for k in range(8):
                ni, nj = i + imove[k], j + jmove[k]
                if 0 <= ni < n and 0 <= nj < m:
                    graph[i*m + j].append(ni*m + nj)
    return graph

def tour(k, v, n, m, graph, mark):
    global count
    if k == n * m:
        count += 1
        mark[v] = k
        print(f"solution #{count}:")
        for i in range(n):
            for j in range(m):
                print(f"{mark[i*m + j]:>2d}", end = " ")
            print()
    else:
        for u in graph[v]:
            if mark[u] == 0:
                mark[u] = k + 1
                tour(k + 1, u, n, m, graph, mark)
                mark[u] = 0

n, m = 3, 4
graph = make_graph(n, m)
for i in range(n * m):
    print(i, ":", graph[i])

total = 0
for s in range(n * m):
    mark = [0] * (n * m)
    mark[s] = 1
    count = 0
    tour(1, s, n, m, graph, mark)
    if count > 0:
        print(s, ":", count)
    total += count
print(f"total = {total}")